packages/flutter/lib/src/material/dropdown_menu_form_field.dart

```diff
@@ -0,0 +1,146 @@
+    Offset? alignmentOffset,
+    FilterCallback<T>? filterCallback,
+    SearchCallback<T>? searchCallback,
+    required List<DropdownMenuEntry<T>> dropdownMenuEntries,
+    List<TextInputFormatter>? inputFormatters,
+    DropdownMenuCloseBehavior closeBehavior = DropdownMenuCloseBehavior.all,
+    super.onSaved,
+    AutovalidateMode? autovalidateMode,
+    super.validator,
+    super.forceErrorText,
+  }) : super(
+         initialValue: initialSelection,
+         autovalidateMode: autovalidateMode ?? AutovalidateMode.disabled,
+         builder: (FormFieldState<T> field) {
+           final _DropdownMenuFormFieldState<T> state = field as _DropdownMenuFormFieldState<T>;
+           void onSelectedHandler(T? value) {
+             field.didChange(value);
+             onSelected?.call(value);
+           }
+
+           return DropdownMenu<T>(
+             enabled: enabled,
+             width: width,
+             menuHeight: menuHeight,
+             leadingIcon: leadingIcon,
+             trailingIcon: trailingIcon,
+             label: label,
+             hintText: hintText,
+             helperText: helperText,
+             errorText: state.errorText,
+             selectedTrailingIcon: selectedTrailingIcon,
+             enableFilter: enableFilter,
+             enableSearch: enableSearch,
+             textStyle: textStyle,
+             textAlign: textAlign,
+             inputDecorationTheme: inputDecorationTheme,
+             menuStyle: menuStyle,
+             controller: controller,
+             initialSelection: state.value,
+             onSelected: onSelectedHandler,
+             requestFocusOnTap: requestFocusOnTap,
+             expandedInsets: expandedInsets,
+             alignmentOffset: alignmentOffset,
+             filterCallback: filterCallback,
+             searchCallback: searchCallback,
+             inputFormatters: inputFormatters,
+             closeBehavior: closeBehavior,
+             dropdownMenuEntries: dropdownMenuEntries,
+           );
+         },
+       );
+
+  /// The callback is called when a selection is made.
+  ///
+  /// Defaults to null. If null, only the text field is updated.
+  final ValueChanged<T?>? onSelected;
+
+  /// Controls the text being edited.
+  ///
+  /// If null, this widget will create its own [TextEditingController].
+  final TextEditingController? controller;
+
+  @override
+  FormFieldState<T> createState() => _DropdownMenuFormFieldState<T>();
+}
+
+class _DropdownMenuFormFieldState<T> extends FormFieldState<T> {
+  DropdownMenuFormField<T> get _dropdownMenuFormField => widget as DropdownMenuFormField<T>;
+
+  @override
+  void didChange(T? value) {
+    super.didChange(value);
+    _dropdownMenuFormField.onSelected?.call(value);
+  }
+
+  @override
+  void didUpdateWidget(DropdownMenuFormField<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.initialValue != widget.initialValue) {
+      setValue(widget.initialValue);
```

Comment:
Should this be done only when the user hasn't selected any other value in the meantime?

Reply:
Good question!
`DropdownButtonFormField` does the same (setting the value to `initialValue` when `initialValue` is changed) but `TextFormField` does not.

Without this logic, one has to call `reset` to get the value back to `initialValue`. With this logic, the 'reset' is done declaratively which seems easier but I agree that there might be use cases where someone wants to change the `initialValue` in case of a future call to `reset` without changing the current edited value.

So it is probably better to remove this logic for this PR.

Reply:
That makes sense to me. If you remove it, then it would be possible for app developers to support both cases, which both seem like reasonable real-world cases to me. Up to you though.

Reply:
> Should this be done only when the user hasn't selected any other value in the meantime?

Implemented in the 4th commit. :rocket: